In [1]:
import numpy as np

In [2]:
def f1(t, x, y):
    return y

def f2(t, x, y):
    return np.sin(t) - x - x**2*y

In [34]:
t = 0
t_target = 1.6
h = 0.04
x = 0
y = 1
X = [x]
Y = [y]
c = [[], []]

In [35]:
for i in range(int(np.round(t_target/h-t/h))):
    c_11 = h*f1(t+i*h, X[i], Y[i])
    c_21 = h*f2(t+i*h, X[i], Y[i])
    
    c_12 = h*f1(t+i*h+h*1/2, X[i]+c_11*1/2, Y[i]+c_21*1/2)
    c_22 = h*f2(t+i*h+h*1/2, X[i]+c_11*1/2, Y[i]+c_21*1/2)
    
    c_13 = h*f1(t+i*h+h*1/2, X[i]+c_12*1/2, Y[i]+c_22*1/2)
    c_23 = h*f2(t+i*h+h*1/2, X[i]+c_12*1/2, Y[i]+c_22*1/2)
    
    c_14 = h*f1(t+i*h+h, X[i]+c_13, Y[i]+c_23)
    c_24 = h*f2(t+i*h+h, X[i]+c_13, Y[i]+c_23)
    
    X.append(X[i]+1/6*(c_11+2*c_12+2*c_13+c_14))
    Y.append(Y[i]+1/6*(c_21+2*c_22+2*c_23+c_24))    

In [36]:
X

[0,
 0.039999785956425955,
 0.0799965605392924,
 0.11998252555169034,
 0.15994459290185536,
 0.19986433715211804,
 0.2397179873964031,
 0.27947646862847253,
 0.3191055022122041,
 0.3585657741411968,
 0.3978131784274955,
 0.43679914115858126,
 0.47547102850574696,
 0.5137726392866373,
 0.5516447796486004,
 0.5890259141549892,
 0.6258528841678697,
 0.6620616811024916,
 0.6975882590761792,
 0.7323693688876581,
 0.7663433933318248,
 0.7994511627401211,
 0.831636729452661,
 0.8628480807301351,
 0.8930377713883463,
 0.9221634601031081,
 0.9501883367405598,
 0.9770814320174072,
 1.0028178050514656,
 1.0273786086747505,
 1.050751036505522,
 1.0729281594948978,
 1.093908662802998,
 1.1136964962969929,
 1.1323004536342867,
 1.1497336957897961,
 1.1660132350483048,
 1.1811593949925778,
 1.195195260985004,
 1.2081461341905384,
 1.2200390004515576]

In [37]:
Y

[1,
 0.9999785625321193,
 0.9998277051318432,
 0.9994160363624861,
 0.9986105734154392,
 0.9972775957185138,
 0.9952837567675157,
 0.9924974445564088,
 0.9887903724149562,
 0.984039372032989,
 0.9781283493917198,
 0.9709503528383053,
 0.9624096913942886,
 0.9524240314753011,
 0.9409263924860862,
 0.927866957203912,
 0.9132146123343305,
 0.8969581387660267,
 0.8791069802039736,
 0.8596915329647045,
 0.8387629182730464,
 0.8163922204606763,
 0.7926691986971678,
 0.7677005046636334,
 0.7416074621534675,
 0.7145234852412413,
 0.6865912279055639,
 0.6579595687047798,
 0.6287805386507745,
 0.5992062987080169,
 0.5693862657891177,
 0.5394644736043142,
 0.5095772384646605,
 0.479851181540982,
 0.4504016395848878,
 0.4213314770676686,
 0.3927302952241269,
 0.3646740184601344,
 0.33722482654542146,
 0.3104313922142081,
 0.2843293782142513]

Step size $h=0.4$ gives $X(1.6) = 1.220254143277014$

Step size $h=0.2$ gives $X(1.6) = 1.2200552932390438$

Step size $h=0.08$ gives $X(1.6) = 1.2200394334624476$

Step size $h=0.04$ gives $X(1.6) = 1.2200390004515576$

The
method of Richardson extrapolation, given in Section 2.3.6, is equally applicable
to problems such as this one involving coupled equations. Since the Runge–
Kutta method has a local error of $O(h)^5$ the global error will be $O(h^4)$. The
method therefore gives the estimated error in the second value of $X(1.6)$ as
$(1.220254 − 1.220055)/15 = 0.000013$. For 6 decimal place accuracy in the final
estimate we need error $≤ 0.0000005$; in other words, the error must be reduced
by a factor of $0.000013/0.0000005 = 26$. Since Runge–Kutta is a fourth-order
method the required step length will be $0.2/\sqrt[4]{26} = 0.088$. Rounding this down to
a suitable size suggests that a step size of $h = 0.08$ will give a solution accurate to
6 decimal places. In fact the program yields, with $h = 0.08$, $X(1.6) = 1.2200394$.
With $h = 0.04$ it gives $X(1.6) = 1.2200390$. Richardson extrapolation predicts the error in the $h = 0.04$ solution as $0.00000003$ and therefore that in the $h = 0.08$ as
$0.0000005$. The required accuracy was therefore just achieved using $h = 0.08$.